### PR TITLE
feat: add advanced encoder options `Config`

### PR DESCRIPTION
**Pull request by duskmoon314**
- Add `convert_case` to dependencies
- Add advanced encoder options `Config`

BREAKING CHANGE
- `Encode` trait needs to implement `encode_with_config`

### CODE REVIEW

**Comment by burrbull** [comment]
Looks good in first view.
Could you also add/modify some tests in `tests` directory?

**Comment by duskmoon314** [comment]
Ok, I will add some tests ASAP.

**Comment by duskmoon314** [comment]
I add some simple tests to show the usage of `svd_encoder::Config`

It's good to add tests for all options but needs more work. orz

File: svd-encoder/src/config.rs [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    /// `APascalCaseExample`
+    Pascal,
+    /// `Snake` case
+    ///
+    /// `a_snake_case_example`
+    Snake,
+    /// `UpperSnake` case
+    ///
+    /// `AN_UPPER_SNAKE_CASE_EXAMPLE`
+    UpperSnake,
```
Comment (burrbull) [inline svd-encoder/src/config.rs:24]: Here http://calebmer.com/inflections/inflections/case/index.html
it is called "constant case"

**Comment by duskmoon314** [comment]
- rename `UpperSnake` to `Constant`
- add tests in all files that are included in `tests/lib.rs`

**Review by burrbull — DISMISSED** [review]

**Comment by burrbull** [comment]
cc @Emilgardis any comments?

File: tests/src/lib.rs [hunk 2]
```diff
@@ -41,6 +41,39 @@ pub fn run_test<
...
+pub fn run_test_with_config<
+    T: Parse<Error = SVDErrorAt, Object = T, Config = Config>
+        + Encode<Error = EncodeError>
+        + core::fmt::Debug
+        + PartialEq,
+>(
+    tests: &[(T, &str, &str)],
+    parser_config: Option<Config>,
+    encoder_config: Option<svd_encoder::Config>,
+) {
```
Comment (Emilgardis) [inline tests/src/lib.rs:53]: I think `run_test` should use this function, so to not duplicate code

**Review by Emilgardis — CHANGES_REQUESTED** [review]
awesome!

missing a optional serialize/deserialize for `Config`, but that's not super important and not a blocker (would just be nice to have to enable some kind of config file in higher level crate).

**Comment by duskmoon314** [comment]
> missing a optional serialize/deserialize

There is no serialize/deserialize for `svd_parser::Config`, so I think it's unnecessary. `svd_encoder::Config` has many fields so it might be beneficial in use. Now I declare an `update` method to set config, and this is the way I use in the temptation of implementing the `format` feature in svdtools.

> I think `run_test` should use this function

Sure. Will fix this soon.

**Comment by duskmoon314** [comment]
Change `run_test` to:

```rust
pub fn run_test<
    T: Parse<Error = SVDErrorAt, Object = T, Config = Config>
        + Encode<Error = EncodeError>
        + core::fmt::Debug
        + PartialEq,
>(
    // tests: [(Object, input string, output string)],
    tests: &[(T, &str, &str)],
    parser_config: Option<Config>,
    encoder_config: Option<svd_encoder::Config>,
)
```

Change all tests using `run_test` (Keep most of them, add several tests for config)

**Comment by Emilgardis** [comment]
> There is no serialize/deserialize for `svd_parser::Config`, so I think it's unnecessary.

Indeed, just saying it's something we should maybe look into :D 

Regarding the `run_test`, I think it's worth having them as two separate functions, and `run_test` calls `runs_test_with_config`.

For that to work `run_test_with_config` should probably take a iterator instead..

However, I'm fine with this.

**Review by Emilgardis — APPROVED** [review]
awesome job, thanks a lot!

one minor issue/non-issue with `run_test`, but it's not a problem

**Comment by duskmoon314** [comment]
> Regarding the `run_test`, I think it's worth having them as two separate functions, and `run_test` calls `runs_test_with_config`.
> 
> For that to work `run_test_with_config` should probably take a iterator instead..

Oh, you are right. That way is more simple to use. I need to think about how to do this.

**Review by burrbull — APPROVED** [review]

**Comment by burrbull** [comment]
Lets merge this. @duskmoon314 if you have more improvements, please open new PR.
bors r+

**Pull request merged**